File: 卡码/代码随想录/1/07-回溯/Python/01-组合/LeetCode_0039.py
#39. 组合总和
from typing import *

class Solution:
  def __init__(self):
    self.path:List[int]=[]
    self.res:List[List[int]]=[]
  def backtracking(self,candidates:List[int],target:int,sum:int,index:int):
    if sum>target:
      return
    elif sum==target:
      self.res.append(self.path[:])
      return
    for i in range(index,len(candidates)):
      if sum+candidates[i]<=target:
        self.path.append(candidates[i])
        self.backtracking(candidates,target,sum+candidates[i],i)
        self.path.pop()
  def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
    candidates.sort()
    self.backtracking(candidates,target,0,0)
    return self.res

File: 卡码/代码随想录/1/07-回溯/Go/01-组合/LeetCode_0039.go
// 39. 组合总和
package main

import (
	"sort"
)

func combinationSum(candidates []int, target int) [][]int {
	path:=make([]int,0)
	res:=make([][]int,0)
	var backtracking func(candidates []int,target,sum,index int)
	backtracking=func(candidates []int,target,sum,index int){
		if sum>target {
			return
		}
		if sum==target {
			tmp:=make([]int,len(path))
			copy(tmp,path)
			res=append(res,tmp)
			return
		}
		for i:=index;i<len(candidates)&&sum+candidates[i]<=target;i++ {
			path=append(path,candidates[i])
			backtracking(candidates,target,sum+candidates[i],i)
			path=path[:len(path)-1]
		}
	}
	sort.Slice(candidates,func(i,j int) bool{
		return candidates[i]<candidates[j]
	})
	backtracking(candidates,target,0,0)
	return res
}
